pydantic/_internal/_typing_extra.py

```diff
@@ -265,24 +265,27 @@ def eval_type_backport(
                 value, globalns, localns
             )
     except TypeError as e:
-        if not (isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)):
-            raise
-        try:
-            from eval_type_backport import eval_type_backport
-        except ImportError:
-            raise TypeError(
-                f'You have a type annotation {value.__forward_arg__!r} '
-                f'which makes use of newer typing features than are supported in your version of Python. '
-                f'To handle this error, you should either remove the use of new syntax '
-                f'or install the `eval_type_backport` package.'
-            ) from e
-
-        return eval_type_backport(value, globalns, localns, try_default=False)
-
-
-def is_backport_fixable_error(e: TypeError) -> bool:
-    msg = str(e)
-    return msg.startswith('unsupported operand type(s) for |: ') or "' object is not subscriptable" in msg
+        if isinstance(value, typing.ForwardRef):
+            try:
+                from eval_type_backport import eval_type_backport
+            except ImportError:
+                raise TypeError(
+                    f'Unable to evaluate type annotation {value.__forward_arg__!r}. If you are making use '
+                    'of the new typing syntax (unions using `|` since Python 3.10 or builtins subscripting '
+                    'since Python 3.9), you should either replace the use of new syntax with the existing '
+                    '`typing` constructs or install the `eval_type_backport` package. Otherwise, it might be '
+                    "that the type being used isn't subscriptable."
+                ) from e
+
+            try:
+                return eval_type_backport(value, globalns, localns, try_default=False)
+            except TypeError as e:
+                raise TypeError(
+                    f'Unable to evaluate type annotation {value.__forward_arg__!r}. '
+                    "It might be that the type being used isn't subscriptable."
```

Comment:
I don't think it makes sense to mention the type not being subscriptable. That should already be in the traceback with the full original error message.

Reply:
Unfortunately only on 3.10+ as I mentioned [here](https://github.com/pydantic/pydantic/pull/10030#discussion_r1701872984) :/

Reply:
So doing something like this on 3.8/3.9:

```python
from __future__ import annotations

class CustomType:
    pass

class Model(BaseModel):
    foo: CustomType[int]
```

is quite hard to debug, as you only get `TypeError: 'type' object is not subscriptable`

Reply:
But you will still have `Unable to evaluate type annotation 'CustomType[int]'`.

Reply:
Ah yes makes sense